File: lib/cs/ProtoPack.cs
using System.Threading;
namespace ILib.ProtoPack
{
	public interface IMessage
	{
		void Write(IWriter w);
		void Read(IReader r);
	}

	public interface IWriter
	{
		int Length { get; }
		byte[] ToArray();
		void Read(byte[] dst, int offest = 0);
		System.ArraySegment<byte> ReadSegment();
		void Clear();
		void WriteMapHeader(int length);
		void WriteArrayHeader(int length);
		void WriteTag(uint tag);
		void WriteNil();
		void Write(double val);
		void Write(float val);
		void Write(long val);
		void Write(ulong val);
		void Write(int val);
		void Write(uint val);
		void Write(bool val);
		void Write(string val);
		void Write(byte[] val);
		void Write(IMessage val);
	}

	public interface IReader
	{
		int Offset { get; }
		void Reset(byte[] buf, int offest = 0);
		void Skip();
		bool NextFormatIsNull();
		uint ReadTag();
		void ReadNil();
		T ReadMessage<T>() where T : IMessage, new();
		byte[] ReadBytes();
		bool ReadBool();
		int ReadInt();
		uint ReadUint();
		long ReadLong();
		ulong ReadUlong();
		double ReadDouble();
		float ReadFloat();
		string ReadString();
		int ReadMapHeader();
		int ReadArrayHeader();
	}

	public static class StaticPacker
	{
		static ThreadLocal<MsgPackWriter> s_Writer = new ThreadLocal<MsgPackWriter>(() => new MsgPackWriter());
		static ThreadLocal<MsgPackReader> s_Reader = new ThreadLocal<MsgPackReader>(() => new MsgPackReader());

		public static byte[] Pack(this IMessage message)
		{
			var w = s_Writer.Value;
			w.Clear();
			message.Write(w);
			return w.ToArray();
		}

		public static T Unpack<T>(this byte[] buf) where T : IMessage, new()
		{
			int offest = 0;
			return Unpack<T>(buf, ref offest);
		}

		public static T Unpack<T>(this byte[] buf, ref int offest) where T : IMessage, new()
		{
			var r = s_Reader.Value;
			r.Reset(buf, offest);
			T message = new T();
			message.Read(r);
			offest = r.Offset;
			r.Reset(null, 0);
			return message;
		}

	}

}

File: lib/golang/protopack.go
package protopack

import (
	"bytes"
	"io"
)

type Message interface {
	Write(w Writer) error
	Read(r Reader) error
}

type Writer interface {
	Length() int
	Bytes() []byte
	Reset()
	WriteMapHeader(length int) error
	WriteArrayHeader(length int) error
	WriteTag(tag uint32) error
	WriteNil() error
	WriteFloat64(val float64) error
	WriteFloat32(val float32) error
	WriteInt64(val int64) error
	WriteUint64(val uint64) error
	WriteInt32(val int32) error
	WriteUint32(val uint32) error
	WriteBool(val bool) error
	WriteString(val string) error
	WriteBytes(val []byte) error
	WriteMessage(val Message) error
}

type Reader interface {
	Reset(r io.Reader)
	Skip() error
	NextFormatIsNull() (bool, error)
	ReadTag() (uint32, error)
	ReadNil() error
	ReadBytes() ([]byte, error)
	ReadBool() (bool, error)
	ReadInt32() (int32, error)
	ReadUint32() (uint32, error)
	ReadInt64() (int64, error)
	ReadUint64() (uint64, error)
	ReadFloat64() (float64, error)
	ReadFloat32() (float32, error)
	ReadString() (string, error)
	ReadMapHeader() (uint, error)
	ReadArrayHeader() (uint, error)
	ReadMessage(msg Message) error
}

func NewWriter(buf bytes.Buffer) Writer {
	return &protoWriter{newMsgWriter(buf)}
}

func NewReader(r io.Reader) Reader {
	return &protoReader{newMsgReader(r)}
}

func Pack(m Message) ([]byte, error) {
	w := NewWriter(bytes.Buffer{})
	if err := m.Write(w); err != nil {
		return nil, err
	}
	return w.Bytes(), nil
}

func Unpack(m Message, buf []byte) error {
	r := NewReader(bytes.NewBuffer(buf))
	return m.Read(r)
}
